libs/core/src/main/java/org/elasticsearch/bootstrap/JavaVersion.java

```diff
@@ -42,23 +45,38 @@ private JavaVersion(List<Integer> version) {
         this.version = Collections.unmodifiableList(version);
     }
 
+    /**
+     * Parses the Java version as it can be retrieved as the value of java.version or
+     * java.specification.version according to JEP 223.
+     *
+     * @param value The version String
+     */
     public static JavaVersion parse(String value) {
         Objects.requireNonNull(value);
+        String prePart = null;
         if (!isValid(value)) {
             throw new IllegalArgumentException("value");
         }
-
         List<Integer> version = new ArrayList<>();
-        String[] components = value.split("\\.");
-        for (String component : components) {
-            version.add(Integer.valueOf(component));
+        String[] parts = value.split("-");
+        String[] numericComponents;
+        if (parts.length == 1) {
+            numericComponents = value.split("\\.");
+        } else if (parts.length == 2) {
+            numericComponents = parts[0].split("\\.");
+            prePart = parts[1];
+        } else {
+            throw new IllegalArgumentException("value");
```

Comment:
We should have a more descriptive failure message, which includes the actual value, not the literal string `value`. ;)